slam-libraries/viam-cartographer/src/slam_service/slam_service.cc

```diff
@@ -92,6 +93,47 @@ ::grpc::Status SLAMServiceImpl::GetPositionNew(
     return grpc::Status::OK;
 }
 
+::grpc::Status SLAMServiceImpl::GetPointCloudMap(
+    ServerContext *context, const GetPointCloudMapRequest *request,
+    GetPointCloudMapResponse *response) {
+    std::string pointcloud_map;
+    // Write or grab the latest pointcloud map in form of a string
+    try {
+        std::shared_lock optimization_lock{optimization_shared_mutex,
+                                           std::defer_lock};
+        if (optimization_lock.try_lock()) {
+            // We are able to lock the optimization_shared_mutex, which means
+            // that the optimization is not ongoing and we can grab the newest
+            // map
+            GetLatestSampledPointCloudMapString(pointcloud_map);
+        } else {
+            // We couldn't lock the mutex which means the optimization process
+            // locked it and we need to use the backed up latest map
+            std::lock_guard<std::mutex> lk(viam_response_mutex);
```

Comment:
Do we want to log in each of these cases so if it turns out that we are frequently returning stale data we have some easy way to detect that?